### PR TITLE
feat(scheduler-service): SFO-58 add free busy API and calendar with subscrip…

### PR DESCRIPTION
**Pull request by IshuSF**
SFO-58, SFO-90

## Description

 - Added free busy api
 - Added calendar with subscription api
 - Updated PUT calendar api to update working hours along with calendar

Fixes # (SFO-58, SFO-90)

## Type of change

Please delete options that are not relevant.

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Intermediate change (work in progress)

## How Has This Been Tested?

Please describe the tests that you ran to verify your changes. Provide instructions so we can reproduce. Please also list any relevant details for your test configuration

- [ ] Via postman and NPM test

## Checklist:

- [x] Performed a self-review of my own code
- [x] npm test passes on your machine
- [x] New tests added or existing tests modified to cover all changes
- [x] Code conforms with the style guide
- [ ] API Documentation in code was updated
- [x] Any dependent changes have been merged and published in downstream modules

### CODE REVIEW

File: services/scheduler-service/src/__tests__/acceptance/calendar.controller.acceptance.ts [hunk 1]
```diff
@@ -94,6 +95,27 @@ describe('Calendar Controller', () => {
     expect(workingHoursResponse.body.dayOfWeek).to.be.equal(1);
   });
 
+  it('add subscription when calendar is added', async () => {
+    const calendarToAdd = {
```
Comment (rohit-sourcefuse) [inline services/scheduler-service/src/__tests__/acceptance/calendar.controller.acceptance.ts:99]:
```
const calendarToAdd = {
      source: 'internal',
      enableWorkingHours: true,
      location: 'location',
      identifier,
      summary: 'string',
      timezone: 'ist',
      subscription: {
        identifier,
        fgColor: "red"
      }
    };
```

File: services/scheduler-service/src/__tests__/acceptance/calendar.controller.acceptance.ts [hunk 2]
```diff
@@ -150,7 +201,7 @@ describe('Calendar Controller', () => {
       source: 'internal',
       enableWorkingHours: true,
       location: 'location',
-      identifier: 'test@gmail.com',
+      identifier: identifier,
```
Comment (rohit-sourcefuse) [inline services/scheduler-service/src/__tests__/acceptance/calendar.controller.acceptance.ts:204]: change like above

File: services/scheduler-service/src/controllers/calendar.controller.ts [hunk 3]
```diff
@@ -59,6 +72,67 @@ export class CalendarController {
...
+
+  @authenticate(STRATEGY.BEARER, {
+    passReqToCallback: true,
+  })
+  @authorize([PermissionKey.CreateCalendar])
+  @post('/calendars/calendarSubscription', {
+    responses: {
+      '200': {
+        description: calendarModelInstance,
+        content: {'application/json': {schema: getModelSchemaRef(CalendarDTO)}},
```
Comment (rohit-sourcefuse) [inline services/scheduler-service/src/controllers/calendar.controller.ts:86]: calendar with subscription.

File: services/scheduler-service/src/controllers/calendar.controller.ts [hunk 4]
```diff
@@ -59,6 +72,67 @@ export class CalendarController {
...
+      identifierType = IdentifierType.Id;
+    }
+    
+    const subscriptionList = await this.subscriptionRepository.find({
+      where: {
+        identifier: this.currentUser[identifierType],
+      },
+    });
+
+    if (subscriptionList.length > 0) {
```
Comment (rohit-sourcefuse) [inline services/scheduler-service/src/controllers/calendar.controller.ts:121]: `subscription.isPrimary = (subscriptionList.length > 0);`

File: services/scheduler-service/src/controllers/calendar.controller.ts [hunk 5]
```diff
@@ -211,9 +285,20 @@ export class CalendarController {
   })
   async replaceById(
     @param.path.string('id') id: string,
-    @requestBody() calendar: Calendar,
+    @requestBody() calendarDTO: CalendarDTO,
   ): Promise<void> {
-    await this.calendarRepository.replaceById(id, calendar);
+    let workingHours: WorkingHour[] = [];
```
Comment (rohit-sourcefuse) [inline services/scheduler-service/src/controllers/calendar.controller.ts:290]: working hours as per discussion.

File: services/scheduler-service/src/controllers/event.controller.ts [hunk 6]
```diff
@@ -108,6 +111,58 @@ export class EventController {
...
+          schema: getModelSchemaRef(FreeBusyDTO, {
+            title: 'FreeBusy request',
+          }),
+        },
+      },
+    })
+    freeBusyDTO: FreeBusyDTO,
+  ) {
+    if (
+      !this.eventService.validateDateForTimeZone(freeBusyDTO.timeMin) ||
```
Comment (rohit-sourcefuse) [inline services/scheduler-service/src/controllers/event.controller.ts:132]: const {timeMin, timeMax} = freeBusyDTO;

File: services/scheduler-service/src/controllers/event.controller.ts [hunk 7]
```diff
@@ -31,6 +31,8 @@ import {
 import {ValidatorService} from '../services/validator.service';
 import {ErrorKeys} from '../models/enums/error-keys';
```
Comment (rohit-sourcefuse) [inline services/scheduler-service/src/controllers/event.controller.ts:32]: Later needed timezone conversation based on parameter.

File: services/scheduler-service/src/models/free-busy-item.dto.ts [hunk 8]
```diff
@@ -0,0 +1,9 @@
+import {property} from '@loopback/repository';
```
Comment (rohit-sourcefuse) [inline services/scheduler-service/src/models/free-busy-item.dto.ts:1]: @IshuSF and @vineet-suri lots of DTO's, so please check we need to call it DTO or interface ?

File: services/scheduler-service/src/controllers/event.controller.ts (hunk 7 above)
Comment (IshuSF) [reply services/scheduler-service/src/controllers/event.controller.ts:32]: need discussion on this. Leaving it for now.

File: services/scheduler-service/src/controllers/calendar.controller.ts [hunk 9]
```diff
@@ -59,25 +88,68 @@ export class CalendarController {
...
-    if (calendarDTO.workingHours) {
-      workingHours = calendarDTO.workingHours;
+    return this.calendarService.createCalendar(calendarDTO);
+  }
+
+  @authenticate(STRATEGY.BEARER, {
+    passReqToCallback: true,
+  })
+  @authorize([PermissionKey.CreateCalendar])
+  @post('/calendars/calendarSubscription', {
```
Comment (vineet-suri) [inline services/scheduler-service/src/controllers/calendar.controller.ts:98]: api url should be in kabab-case.

File: services/scheduler-service/src/controllers/event.controller.ts [hunk 10]
```diff
@@ -108,6 +111,56 @@ export class EventController {
     return event;
   }
 
+  @authenticate(STRATEGY.BEARER, {
+    passReqToCallback: true,
+  })
+  @authorize([PermissionKey.ViewEvent, PermissionKey.ViewAttendee])
+  @get('/events/freeBusy')
```
Comment (vineet-suri) [inline services/scheduler-service/src/controllers/event.controller.ts:118]: url in kabab-case.

**Pull request merged**